cloud,backupccl: add metric for time of last backup failure due to KMS

rhu713 opened the pull request:
Add a metric `backup.last-failed-time.kms-inaccessible` that tracks the
timestamp of the last backup failure due to a KMS error. This metric is only
tracked for BACKUP statrements that are executed with the OPTION
`updates_cluster_monitoring_metrics` set.

Release note (sql change): adds the `updates_cluster_monitoring_metrics` backup
option that allows an operator to opt-in tracking the timestamp
of the last backup failure due to a KMS error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104634)
<!-- Reviewable:end -->

Commit:
cloud,backupccl: add metric for time of last backup failure due to KMS

Add a metric `backup.last-failed-time.kms-inaccessible` that tracks the
timestamp of the last backup failure due to a KMS error. This metric is only
tracked for BACKUP statrements that are executed with the OPTION
`updates_cluster_monitoring_metrics` set.

Release note (sql change): adds the `updates_cluster_monitoring_metrics` backup
option that allows an operator to opt-in tracking the timestamp
of the last backup failure due to a KMS error.

msbutler commented on pkg/ccl/backupccl/create_scheduled_backup.go:
```diff
@@ -291,6 +293,17 @@ func doCreateBackupSchedules(
 
+	_, updatesErrorMetrics := scheduleOptions[optUpdatesErrorMetrics]
+
+	if updatesErrorMetrics {
```
what's the motivation behind adding a backup schedule option for this change?

rhu713 replied:
this feature is something that we want to be opt-in so that e.g. customer created schedules don't write this metric by default. This is in line with the similar RPO metric that's controlled by `optUpdatesLastBackupMetric`

msbutler replied:
I see, that clarifies things. thanks! It may be worth adding one sentence explaining the opt-in motivation in the commit message or in the code, namely: " customer created schedules don't write this metric by default."

msbutler commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -1156,9 +1156,7 @@ func (b *backupResumer) maybeNotifyScheduledJobCompletion(
-		if err := jobs.NotifyJobTermination(
-			ctx, txn, env, b.job.ID(), jobStatus, b.job.Details(), scheduleID,
-		); err != nil {
+		if err := jobs.NotifyJobTermination(ctx, txn, env, b.job.ID(), jobStatus, jobErr, b.job.Details(), scheduleID); err != nil {
```
instead of refactoring the `NotifyJobTermination` interface, why not increment the metric here, before we call `NotifyJobTermination`?

rhu713 replied:
I think in this case we'd need to pass in the same bool that tells the job whether or not it should increment the KMS into the backup job. But I also think that it's not a bad idea to have `jobs.NotifyJobTermination` to notify with the error either since there may be other use cases for this in the future.

rhu713 replied:
I moved the metric to the BACKUP statement instead. PTAL

msbutler commented on pkg/ccl/backupccl/schedule_exec.go:
```diff
@@ -164,23 +166,25 @@ func (e *scheduledBackupExecutor) NotifyJobTermination(
 ) error {
 	if jobStatus == jobs.StatusSucceeded {
 		e.metrics.NumSucceeded.Inc(1)
-		log.Infof(ctx, "backup job %d scheduled by %d succeeded", jobID, schedule.ScheduleID())
```
this implementation change makes this function a bit harder to understand: we return an error to the user, but don't use the newly passed in the `jobErr`. I wonder if we can avoid this big change by merely incrementing this metric in the backup job itself, as I propose.

rhu713 replied:
same as before, I think it's good to notify with the error anyways since it's something that could be useful for jobs with a failed status.

msbutler replied:
hrm: I wonder if there's a way to add this feature without complicating the error handling here. It doesn't sit well with me that we have `jobErr` passed in and a different error generated from the `jobStatus` returned to the user. I'm going to ping job peeps to take a look at this.

dt reviewed (commented):
This this could potentially benefit from being broken up into a few commits that would stand on their own: tag kms errors with a marker, cleanup a few kms errors, add err to notify API, then finally add a metric updated on kms-related errors.

msbutler commented on pkg/ccl/backupccl/backup_metrics.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// MetricStruct implements the metric.Struct interface.
+func (b BackupMetrics) MetricStruct() {}
```
this is great. will make it much easier to add new metrics when we need to

msbutler reviewed (approved):
this looks great!! thanks for being patient with me and david.

rhu713 commented:
bors r+

rhu713 commented:
bors r+

adityamaru commented:
does this need a dev gen? - https://teamcity.cockroachdb.com/viewLog.html?buildId=11857186&buildTypeId=Cockroach_BazelEssentialCi

yuzefovich commented:
Needs `./dev gen docs`

bors r-

rhu713 commented:
bors r+
